### ISSUE TITLE
Tracing (Edinburgh Roadmap Item)

### ISSUE THREAD

**Issue by jpwhitemn** [edgexfoundry/edgex-go#824]
Tracing is a stretch goal task for the Edinburgh release.  Tracing is the ability to trace a system API request inside and across EdgeX services for the purpose of debugging and performance metrics tracking.  Tracing is provided by frameworks such as GoKit, but EdgeX has not yet adopted such a framework and needs a solution that is language independent (for use across non-Go services like C based Device Services).

**Comment by tsconn23** [comment]
Most likely for Edinburgh we will have to be satisfied with a "poor man's solution" to tracing. This is predicated on the inclusion of the GoKit style logging current pending in PR #787. Simply put, every request handler will need to do the following:

1.) Interrogate HTTP headers on the incoming request to see if the "correlation-id" header is present. If not, create a correlation-id value (should be UUID format)
2.) Put the correlation-id value into the current context
3.) If a call to another service is made (like core-data to core-metadata) the correlation-id should be pulled from the context and its value used to set the header on the outgoing request
4.) When the request handler completes, regardless of success or failure, the correlation-id must be logged via the GoKit logger with its own key/value.

**Comment by tsconn23** [comment]
@SteveOss @cloudxxx8 Please note, there's a device service work item here in that every call to the core services will need to contain a custom HTTP header

"correlation-id":<UUID value>

**Comment by cloudxxx8** [comment]
@tsconn23 
Since device-sdk-go is calling LogginClient to send requests, I thought we need to add an optional parameter to pass correlation-id and put the "X-Correlation-ID HTTP header" logic into LoggingClient.  What do you think?

**Comment by tsconn23** [comment]
@cloudxxx8 Well, yes in the case where the support-logging service is involved. But every client needs this same capability.

**Comment by cloudxxx8** [comment]
@tsconn23 
You are right.  I will make device-sdk-go supports it once this capability of each clients is ready.

**Comment by tsconn23** [comment]
Closing. Replaced by issue #1014. Work is being done in PR #1067.
